source/Mlos.Model.Services.Client/OptimizerServiceEncoderDecoder.cs

```diff
@@ -0,0 +1,419 @@
+// -----------------------------------------------------------------------
+// <copyright file="OptimizerServiceEncoderDecoder.cs" company="Microsoft Corporation">
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root
+// for license information.
+// </copyright>
+// -----------------------------------------------------------------------
+
+using Mlos.Model.Services.Spaces;
+
+namespace Mlos.Model.Services.Client
+{
+    public class OptimizerServiceEncoder
+    {
+        public OptimizerService.SimpleHypergrid EncodeHypergrid(Hypergrid hypergrid)
+        {
+            OptimizerService.SimpleHypergrid instance = new OptimizerService.SimpleHypergrid();
+
+            // TODO: I have no idea if this is the right way to handle subgrids.
+            foreach (var subgridSet in hypergrid.Subgrids.Values)
+            {
+                foreach (var subgrid in subgridSet)
+                {
+                    instance.GuestSubgrids.Add(EncodeSubgridJoin(subgrid));
+                }
+            }
+
+            foreach (var dimension in hypergrid.Dimensions)
+            {
+                instance.Dimensions.Add(EncodeDimension(dimension));
+            }
+
+            instance.Name = hypergrid.Name;
+
+            return instance;
+        }
+
+        public OptimizerService.OptimizationProblem EncodeOptimizationProblem(OptimizationProblem problem)
+        {
+            OptimizerService.OptimizationProblem instance = new OptimizerService.OptimizationProblem();
+            instance.ParameterSpace = EncodeHypergrid(problem.ParameterSpace);
+            instance.ObjectiveSpace = EncodeHypergrid(problem.ObjectiveSpace);
+            foreach (var objective in problem.Objectives)

```

Comment:
try to use Linq instead foreach

Reply:
[zaporter/grpc dd30df83] Updated to use Linq in the EncoderDecoder. Also changed Smartcache to use a primitive value rather than enum.